File: scripts/list_colors.py
import re
import os
from collections import Counter

ROOT = os.path.join(os.path.dirname(__file__), "..", "src")

hex_pat = re.compile(r"#([0-9a-fA-F]{3,8})\b")
rgba_pat = re.compile(
    r"rgba?\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)(?:\s*,\s*([\d.]+))?\s*\)"
)


def norm_hex(h: str) -> str:
    h = h.lower()
    if len(h) == 3:
        return "#" + "".join(c * 2 for c in h)
    if len(h) == 4:
        return "#" + "".join(a + b for a, b in zip(h[0::2], h[1::2]))
    if len(h) == 6:
        return "#" + h
    if len(h) == 8:
        return "#" + h[:6]
    return "#" + h


hexes: Counter[str] = Counter()
rgbas: Counter[str] = Counter()
by_file: dict[str, set[str]] = {}

for root, dirs, files in os.walk(ROOT):
    dirs[:] = [d for d in dirs if d not in ("node_modules",)]
    for name in files:
        if not name.endswith((".css", ".tsx", ".ts", ".jsx", ".js")):
            continue
        path = os.path.join(root, name)
        rel = os.path.relpath(path, ROOT)
        try:
            text = open(path, encoding="utf-8", errors="ignore").read()
        except OSError:
            continue
        fh: set[str] = set()
        for m in hex_pat.finditer(text):
            h = norm_hex(m.group(1))
            hexes[h] += 1
            fh.add(h)
        for m in rgba_pat.finditer(text):
            a = m.group(4) or "1"
            r = f"rgba({m.group(1)}, {m.group(2)}, {m.group(3)}, {a})"
            rgbas[r] += 1
        if fh:
            by_file[rel] = fh

print(f"HEX unique: {len(hexes)}, refs: {sum(hexes.values())}")
for h, c in sorted(hexes.items(), key=lambda x: (-x[1], x[0])):
    print(f"{c:5d}  {h}")

print(f"\nRGBA/RGB unique: {len(rgbas)}, refs: {sum(rgbas.values())}")
for r, c in sorted(rgbas.items(), key=lambda x: -x[1])[:50]:
    print(f"{c:5d}  {r}")

print("\nNon-App.css hex in TS/TSX:")
for rel, colors in sorted(by_file.items()):
    if rel == "App.css":
        continue
    for h in sorted(colors):
        print(f"  {rel}: {h}")

File: scripts/list-colors.mjs
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const ROOT = path.join(path.dirname(fileURLToPath(import.meta.url)), "../src");
const hexRe = /#([0-9a-fA-F]{3,8})\b/g;
const rgbaRe = /rgba?\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)(?:\s*,\s*([\d.]+))?\s*\)/g;

function normHex(h) {
  h = h.toLowerCase();
  if (h.length === 3) return "#" + [...h].map((c) => c + c).join("");
  if (h.length === 6) return "#" + h;
  if (h.length === 8) return "#" + h.slice(0, 6);
  return "#" + h;
}

const hex = new Map();
const rgba = new Map();
const otherFiles = new Map();

function walk(dir) {
  for (const name of fs.readdirSync(dir)) {
    if (name === "node_modules") continue;
    const p = path.join(dir, name);
    const st = fs.statSync(p);
    if (st.isDirectory()) {
      walk(p);
      continue;
    }
    if (!/\.(css|tsx?|jsx?)$/.test(name)) continue;
    const text = fs.readFileSync(p, "utf8");
    const rel = path.relative(ROOT, p);
    let m;
    hexRe.lastIndex = 0;
    while ((m = hexRe.exec(text))) {
      const h = normHex(m[1]);
      hex.set(h, (hex.get(h) || 0) + 1);
      if (rel !== "App.css") {
        if (!otherFiles.has(rel)) otherFiles.set(rel, new Set());
        otherFiles.get(rel).add(h);
      }
    }
    rgbaRe.lastIndex = 0;
    while ((m = rgbaRe.exec(text))) {
      const r = `rgba(${m[1]}, ${m[2]}, ${m[3]}, ${m[4] ?? "1"})`;
      rgba.set(r, (rgba.get(r) || 0) + 1);
    }
  }
}

walk(ROOT);

const sortEntries = (map) =>
  [...map.entries()].sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]));

console.log(`HEX unique: ${hex.size}, refs: ${[...hex.values()].reduce((a, b) => a + b, 0)}`);
for (const [h, c] of sortEntries(hex)) {
  console.log(`${String(c).padStart(5)}  ${h}`);
}

console.log(`\nRGBA unique: ${rgba.size}, refs: ${[...rgba.values()].reduce((a, b) => a + b, 0)}`);
for (const [r, c] of sortEntries(rgba).slice(0, 35)) {
  console.log(`${String(c).padStart(5)}  ${r}`);
}

console.log("\nHex in non-App.css files:");
for (const [f, set] of [...otherFiles.entries()].sort()) {
  console.log(`  ${f}: ${[...set].sort().join(", ")}`);
}
